Add support for nested error element and no error element in SimpleForm config

TeWu opened the pull request:
Field wrappers in _SimpleForm_ config can have multiple nested elements (wrappers), so error element don't have to be in the topmost field wrapper e.g.:

``` ruby
config.wrappers :my_custom_wrapper, tag: :div, class: :field do |b|
  # Form extensions
  b.use :html5

  # Form components
  b.use :label
  b.wrapper tag: :div, class: :input do |bi|
    bi.use :input
    bi.use :error, wrap_with: {tag: :span, class: :err}
  end
end
```

Current implementation of _ClientSideValidations-SimpleForm_ always adds error tag in the topmost field wrapper (in above example `div.field`), and doesn't determine position of error element as it is specified in _SimpleForm_ config. Even when error element is not present in _SimpleForm_ config, the error tag will be added as long as form has `data-validate` attribute set to `true`.

This PR makes _ClientSideValidations-SimpleForm_ first determine position of error element as it is specified in appropriate wrapper definition in _SimpleForm_ config, and than add error tag in the proper position. This also fixes issue #54.

tagliala commented on coffeescript/rails.validations.simple_form.coffee:
```diff
@@ -4,6 +4,9 @@
 ###
 
+join_tag_and_classes = (tag, classes) ->
+  tag + '.' + classes.filter(Boolean).map((cls) -> cls.replace(/\ /g, '.')).join('.')
```
I really like this approach but we cannot use this

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map

map and filter requires IE9 but the current versions of CSV still supports IE8

TeWu commented:
OK - So I was a little too quick with the statement about this PR fixing issue #54. But I've added some commits (making PR even more complex ;)), and now this PR should fix issue #54. This time I've added some tests to back this statement.

jclusso commented:
Any updates on this? I also have our errors nested so they show up next to the label and without this I can't use the gem :(

tagliala commented:
Hi, please take a look here https://github.com/DavyJonesLocker/client_side_validations-simple_form/pull/56#r69717418

jclusso commented:
I saw that.... but it was like 4 months ago. Wasn't sure if anything had changed.

TeWu commented:
Why not go with Polyfills, as I already [proposed][1]?

Alternatively, I could rewrite `join_tag_and_classes` to avoid using `filter` and `map`, but I think Polyfills are better, because they can be easily deleted after IE8 support is dropped.

[1]: https://github.com/DavyJonesLocker/client_side_validations-simple_form/pull/56#issuecomment-230823522

tagliala commented:
We can use jQuery's `map` and `filter`

jclusso commented:
@TeWu does your branch work in everything but IE8. Because I was still getting an error using your fork.

```ruby
config.wrappers :default, class: 'input-wrapper', hint_class: :hint, error_class: :error do |b|
    b.use :html5
    b.use :placeholder
    b.optional :maxlength
    b.optional :pattern
    b.optional :min_max
    b.optional :readonly

    b.wrapper tag: :label, class: 'field-wrap' do |fw|
      fw.wrapper tag: :div, class: 'label' do |l|
        l.use :label_text
        l.use :tooltip
        l.use :error
        l.use :label_link
      end
      fw.wrapper tag: :div, class: 'field' do |f|
        f.use :input
      end
    end
  end```

TeWu commented:
@jclusso - You need to change

    l.use :error

to

    l.use :error, wrap_with: {tag: :span, class: :error}

tagliala commented:
Thanks for this PR.

It required without any doubt a lot of effort on your side and so I'm sorry to close it.

Could you please update changes to 5.0.0? I did a lot of changes

Please also try to tell me if this is a breaking change of it isn't  because I need to know if I must bump the major version number or not

Also, `join_tag_and_classes` should use `camelCase`. There was a pending review for months, my mistake, the new interface confuses me
